File: Server-Go-Fiber/entity/page.go
package entity

import (
	"errors"
	"strconv"
	"database/sql"
	_ "github.com/mattn/go-sqlite3"
)

type Page struct {
	Id     int  `json:"id"`
	Url  string  `json:"url"`
	Title  string  `json:"title"`
	Content string  `json:"content"`
	// DatePublished sql.NullInt64  `json:"date_published"`
	// DateModified sql.NullInt64  `json:"date_modified"`
	DatePublished *int  `json:"date_published"`
	DateModified *int  `json:"date_modified"` // Prevent: converting NULL to int is unsupported
	UserId  int `json:"user_id"`
}

func GetAll() ([]Page, error) {
	db, err := connect() // connect() declared in user.go
    if err != nil {
        return []Page{}, err
    }
    defer db.Close()

	rows, err := db.Query("SELECT * FROM page")
	if err != nil {
		return nil, err
	}
 
	pages := []Page{}
	for rows.Next() {
		var page Page
		if err := rows.Scan(&page.Id, &page.Url, &page.Title, &page.Content, &page.DatePublished, &page.DateModified, &page.UserId); err != nil {
			return nil, err
		}
		pages = append(pages, page)
	}
	return pages, nil
}

func GetByIdOrUrl(id string) (Page, error) {
	db, err := connect()
	if err != nil {
		return Page{}, err
	}
	defer db.Close()

	var page Page
	var row *sql.Row
	if i, err := strconv.Atoi(id); err == nil {
		// Get page by id
		row = db.QueryRow("SELECT * FROM page WHERE id = ?", i)
	} else {
		// Get page by url
		row = db.QueryRow("SELECT * FROM page WHERE url = ?", id)
	}
	if err := row.Scan(&page.Id, &page.Url, &page.Title, &page.Content, &page.DatePublished, &page.DateModified, &page.UserId); err != nil {
		if err == sql.ErrNoRows {
			return Page{}, errors.New("Page not found: " + id)
		} else {
			return Page{}, err
		}
	}
	return page, nil
}

func Post(page Page) (error) {
	db, err := connect()
	if err != nil {
		return err
	}
	defer db.Close()

    _, err = db.Exec("INSERT INTO page (url, title, content, date_published, date_modified, user_id) VALUES (?, ?, ?, ?, ?, ?)", 
										page.Url, page.Title, page.Content, page.DatePublished, page.DateModified, page.UserId)
    if err != nil {
        return err
    }
    // id, err := result.LastInsertId()
    // if err != nil {
    //     return err
    // }
    return nil
}

func Put(page Page) (error) {
	db, err := connect()
	if err != nil {
		return err
	}
	defer db.Close()

	// _, err = db.Exec("UPDATE page SET url = ?, title = ?, content = ?, date_published = ?, date_modified = ?, user_id = ? WHERE id = ?",
	// 									page.Url, page.Title, page.Content, page.DatePublished, page.DateModified, page.UserId, page.Id)
	sql := "UPDATE page SET "
	if page.Url != "" {
		sql += "url = '" + page.Url + "', "
	}
	if page.Title != "" {
		sql += "title = '" + page.Title + "', "
	}
	if page.Content != "" {
		sql += "content = '" + page.Content + "', "
	}
	if page.DatePublished != nil {
		sql += "date_published = " + strconv.Itoa(*page.DatePublished) + ", "
	}
	if page.DateModified != nil {
		sql += "date_modified = " + strconv.Itoa(*page.DateModified) + ", "
	}
	if page.UserId != 0 {
		sql += "user_id = " + strconv.Itoa(page.UserId) + ", "
	}
	sql = sql[:len(sql)-2] + " WHERE id = " + strconv.Itoa(page.Id)
	_, err = db.Exec(sql)
	if err != nil {
		return err
	}
	return nil
}

func Delete(id int) (error) {
	db, err := connect()
	if err != nil {
		return err
	}
	defer db.Close()

	_, err = db.Exec("DELETE FROM page WHERE id = ?", id)
	if err != nil {
		return err
	}
	return nil
}

File: Server-Rust-Axum/src/entities/page.rs
use serde::{Deserialize, Serialize};
use sqlx::{prelude::FromRow, sqlite::SqliteQueryResult, Error, Pool, Sqlite};

#[derive(FromRow, Serialize, Deserialize)]
pub struct Page {
    id: u32,
    url: String,
    title: String,
    content: String,
    date_published: i64,
    date_modified: i64,
    user_id: u32,
}

impl Page {
    pub fn new(
        url: String,
        title: String,
        content: String,
        date_published: i64,
        date_modified: i64,
        user_id: u32,
    ) -> Self {
        Self {
            id: 0,
            url,
            title,
            content,
            date_published,
            date_modified,
            user_id,
        }
    }

    pub fn new_with_id(
        id: u32,
        url: String,
        title: String,
        content: String,
        date_published: i64,
        date_modified: i64,
        user_id: u32,
    ) -> Self {
        Self {
            id,
            url,
            title,
            content,
            date_published,
            date_modified,
            user_id,
        }
    }

    pub fn get_author_id(&self) -> u32 {
        self.user_id
    }
}

pub async fn get_all(pool: &Pool<Sqlite>) -> Result<Vec<Page>, Error> {
    let sql = "SELECT * FROM page".to_owned();
    let result = sqlx::query_as::<_, Page>(&sql).fetch_all(pool).await;
    result
}

pub async fn get_by_id_or_url(pool: &Pool<Sqlite>, id: String) -> Result<Page, Error> {
    let result = id.parse::<u32>();
    match result {
        Ok(id) => get_by_id(pool, id).await,
        Err(_) => get_by_url(pool, id).await,
    }
}

pub async fn get_by_id(pool: &Pool<Sqlite>, id: u32) -> Result<Page, Error> {
    let sql = format!("SELECT * FROM page WHERE id = {}", id);
    let result = sqlx::query_as::<_, Page>(&sql).fetch_one(pool).await;
    result
}

pub async fn get_by_url(pool: &Pool<Sqlite>, url: String) -> Result<Page, Error> {
    let sql = format!("SELECT * FROM page WHERE url = '{}'", url);
    let result = sqlx::query_as::<_, Page>(&sql).fetch_one(pool).await;
    result
}

pub async fn post(pool: &Pool<Sqlite>, page: Page) -> Result<SqliteQueryResult, Error> {
    let sql = format!("INSERT INTO page (url, title, content, date_published, date_modified, user_id) VALUES ('{}', '{}', '{}', {}, {}, {})",
            page.url,
            page.title,
            page.content,
            page.date_published,
            page.date_modified,
            page.user_id,
    );
    let result = sqlx::query(&sql).execute(pool).await;
    result
}

pub async fn put(pool: &Pool<Sqlite>, page: Page) -> Result<SqliteQueryResult, Error> {
    let sql = format!("UPDATE page SET url='{}', title='{}', content='{}', date_published={}, date_modified={}, user_id={} WHERE id={}", 
            page.url,
            page.title,
            page.content,
            page.date_published,
            page.date_modified,
            page.user_id,
            page.id,
    );
    let result = sqlx::query(&sql).execute(pool).await;
    result
}

pub async fn delete(pool: &Pool<Sqlite>, id: u32) -> Result<SqliteQueryResult, Error> {
    let sql = format!("DELETE FROM page WHERE id={}", id);
    let result = sqlx::query(&sql).execute(pool).await;
    result
}
